aten/src/ATen/native/cuda/TensorTransformations.cu

```diff
@@ -0,0 +1,163 @@
+#include "ATen/NativeFunctions.h"
+#include "ATen/ATen.h"
+#include <algorithm>
+#include <sstream>
+
+#include "ATen/cuda/AccumulateType.cuh"
+#include "ATen/cuda/CUDATensorMethods.cuh"
+#include "ATen/cuda/CUDATypeConversion.cuh"
+
+
+namespace at {
+namespace native {
+
+// Map the index of an element in tensor from 1D to nD
+__device__ __forceinline__
+void linear_index_to_indices(int64_t linear_index, int64_t* each_dim_len, int64_t total_dims, int64_t* indices) {
+  int64_t res = linear_index;
+  for (int64_t i = 0; i < total_dims; i++) {
+    int64_t indices_i = linear_index * total_dims + i;
+    indices[indices_i] = res / each_dim_len[i];
+    res = res % each_dim_len[i];
+  }
+}
+
+/*
+Map the index of an element in tensor from nD to 1D. A tensor is originally in nD shape,
+and 1D is the unfolded version of it (a vector).
+
+Example: given a 3D tensor
+[
+  [ [1,2], [3,4] ],
+  [ [5,6], [7,8] ]
+]
+
+Here element 3 has nD index = (0,1,0), and this corresponds to oneD index = 2
+*/
+__device__ __forceinline__
```

Comment:
I’m having trouble understanding the code and how it relates to the example. Why is there a src_linear_index?

Reply:
src_linear_index is the linear index of an element before flip. I used "src" here as in contrast to "dest_linear_index", which is the index of an element after flip. I add a bit more explanation for this function.